### PR TITLE
Add new HTTPConnectionPool Manager

### PR DESCRIPTION
**Pull request by fabianfett**
Follow up to #418

### CODE REVIEW

File: Sources/AsyncHTTPClient/ConnectionPool/HTTPConnectionPool+Manager.swift [hunk 1]
```diff
@@ -12,7 +12,157 @@
...
+import Logging
+import NIO
 import NIOConcurrencyHelpers
+import NIOHTTP1
+
+extension HTTPConnectionPool {
+    final class Manager {
+        private typealias Key = ConnectionPool.Key
+
+        enum State {
```
Comment (glbrntt) [inline Sources/AsyncHTTPClient/ConnectionPool/HTTPConnectionPool+Manager.swift:24]: nit: private

File: Sources/AsyncHTTPClient/ConnectionPool/HTTPConnectionPool+Manager.swift [hunk 2]
```diff
@@ -12,7 +12,157 @@
...
+        init(eventLoopGroup: EventLoopGroup,
+             configuration: HTTPClient.Configuration,
+             backgroundActivityLogger logger: Logger) {
+            self.eventLoopGroup = eventLoopGroup
+            self.configuration = configuration
+            self.logger = logger
+        }
+
+        deinit {
+            guard case .shutDown = self.state else {
```
Comment (glbrntt) [inline Sources/AsyncHTTPClient/ConnectionPool/HTTPConnectionPool+Manager.swift:50]: What about shutting down? You could call shutdown and not bother waiting for the future to complete as well, I assume?

File: Sources/AsyncHTTPClient/ConnectionPool/HTTPConnectionPool+Manager.swift [hunk 3]
```diff
@@ -12,7 +12,157 @@
...
+
+            switch poolResult {
+            case .success(let pool):
+                pool.executeRequest(request)
+            case .failure(let error):
+                request.fail(error)
+            }
+        }
+
+        func shutdown(on eventLoop: EventLoop) -> EventLoopFuture<Bool> {
```
Comment (glbrntt) [inline Sources/AsyncHTTPClient/ConnectionPool/HTTPConnectionPool+Manager.swift:90]: What's the `Bool` here?

File: Sources/AsyncHTTPClient/ConnectionPool/HTTPConnectionPool+Manager.swift [hunk 4]
```diff
@@ -12,7 +12,157 @@
...
+            let action = self.lock.withLock { () -> ShutdownAction in
+                switch self.state {
+                case .active:
+                    // If there aren't any pools, we can mark the pool as shut down right away.
+                    if self._pools.isEmpty {
+                        let future = eventLoop.makeSucceededFuture(true)
+                        self.state = .shutDown
+                        return .done(future)
+                    } else {
+                        let promise = eventLoop.makePromise(of: Bool.self)
```
Comment (glbrntt) [inline Sources/AsyncHTTPClient/ConnectionPool/HTTPConnectionPool+Manager.swift:105]: Worth a comment here to indicate when this promise is completed.

File: Sources/AsyncHTTPClient/ConnectionPool/HTTPConnectionPool+Manager.swift [hunk 5]
```diff
@@ -12,7 +12,157 @@
...
+        }
+
+        let closeAction = self.lock.withLock { () -> CloseAction in
+            switch self.state {
+            case .active, .shutDown:
+                preconditionFailure("Why are pools shutting down, if the manager did not give a signal")
+
+            case .shuttingDown(let promise, let soFarUnclean):
+                guard self._pools.removeValue(forKey: pool.key) === pool else {
+                    preconditionFailure("Expected that the pool was ")
```
Comment (glbrntt) [inline Sources/AsyncHTTPClient/ConnectionPool/HTTPConnectionPool+Manager.swift:145]: That suspense is killing me!

File: Sources/AsyncHTTPClient/ConnectionPool/HTTPConnectionPool+Manager.swift [hunk 6]
```diff
@@ -12,7 +12,157 @@
...
+        enum State {
+            case active
+            case shuttingDown(promise: EventLoopPromise<Bool>, unclean: Bool)
+            case shutDown
+        }
+
+        let eventLoopGroup: EventLoopGroup
+        let configuration: HTTPClient.Configuration
+        let connectionIDGenerator = Connection.ID.globalGenerator
+        let logger: Logger
```
Comment (glbrntt) [inline Sources/AsyncHTTPClient/ConnectionPool/HTTPConnectionPool+Manager.swift:33]: private?

**Review by Lukasa — APPROVED** [review]

**Review by glbrntt — APPROVED** [review]
LGTM

**Pull request merged**